File: 0002_Add_Two_Numbers/python3/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        d1 = l1
        d2 = l2
        head = ListNode(-1)
        current = head
        carry = 0

        # O(n): Iterate through each digit node.

        while True:
            if not d1 and not d2:
                break
            elif d1 and d2:
                s = d1.val + d2.val + carry
                d1 = d1.next
                d2 = d2.next
            elif d1 and not d2:
                s = d1.val + carry
                d1 = d1.next
            else: #not d1 and d2:
                s = d2.val + carry
                d2 = d2.next
            
            if s > 9:
                carry = 1
                s -= 10
            else:
                carry = 0
            current.next = ListNode(s)
            current = current.next
        
        if carry: current.next = ListNode(carry)
        return head.next

File: 0002_Add_Two_Numbers/go/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    // O(n): Iterate over each digit one after the other
    node1 := l1
    node2 := l2
    head := &ListNode{Val: -1, Next: nil}
    current := head
    carry := 0
    sum := 0

    for true {
        if node1 == nil && node2 == nil {
            break
        } else if node1 != nil && node2 != nil {
            sum = node1.Val + node2.Val + carry
            node1 = node1.Next
            node2 = node2.Next
        } else if node1 != nil && node2 == nil {
            sum = node1.Val + carry
            node1 = node1.Next
        } else {
            sum = node2.Val + carry
            node2 = node2.Next
        }

        if sum >= 10 {
            sum -= 10
            carry = 1
        } else {
            carry = 0
        }

        current.Next = &ListNode{Val: sum, Next: nil}
        current = current.Next
    }

    if carry == 1 {
        current.Next = &ListNode{Val: carry, Next: nil}
    }
    return head.Next
}
